### PR TITLE
Implementation of new Batching API

### PR DESCRIPTION
**Pull request by rahulKQL**

### CODE REVIEW

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 1]
```diff
@@ -886,7 +889,10 @@ public void mutateRow(RowMutation rowMutation) {
    *   // After `batcher` is closed, all mutations have been applied
    * }
    * }</pre>
+   *
+   * @deprecated Please use latest {@link #newBulkMutationBatcher(String)} API.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: Please remove the other docs and only retain the deprecation notice

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java [hunk 2]
```diff
@@ -100,4 +100,10 @@ public MutateRowsRequest toProto(RequestContext requestContext) {
         .setAppProfileId(requestContext.getAppProfileId())
         .build();
   }
+
+  public BulkMutation clone() {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java:116]: `@Override`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutationBatcher.java [hunk 3]
```diff
@@ -33,7 +33,11 @@
  *
  * @see com.google.cloud.bigtable.data.v2.BigtableDataClient#newBulkMutationBatcher() for example
  *     usage.
+ *
+ * @deprecated @deprecated Please use {@link
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutationBatcher.java]: - please remove the 2nd deprecated
- also please remove all other docs other then deprecation notice

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+  @Override
+  public RowMutationEntry setCell(
+      @Nonnull String familyName, @Nonnull String qualifier, @Nonnull String value) {
+    mutation.setCell(familyName, qualifier, value);
+    return this;
+  }
+
+  @Override
+  public RowMutationEntry setCell(
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java:67]: Please inherit javadoc

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 5]
```diff
@@ -389,7 +389,8 @@ public Builder toBuilder() {
     private final UnaryCallSettings.Builder<Query, Row> readRowSettings;
     private final UnaryCallSettings.Builder<String, List<KeyOffset>> sampleRowKeysSettings;
     private final UnaryCallSettings.Builder<RowMutation, Void> mutateRowSettings;
-    private final BatchingCallSettings.Builder<RowMutation, Void> bulkMutateRowsSettings;
+    private final BatchingCallSettings.Builder<RowMutationEntry, Void, RowMutation, Void>
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Why is the request type RowMutation? Shouldn't it be BulkMutation?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 6]
```diff
@@ -501,36 +502,26 @@ private Builder(EnhancedBigtableStubSettings settings) {
      * constructing the callable chain.
      */
     private static class PlaceholderBatchingDescriptor
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: I don't think the placeholder should be necessary anymore

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 7]
```diff
@@ -75,6 +79,8 @@
 public class EnhancedBigtableStub implements AutoCloseable {
   private static final String TRACING_OUTER_CLIENT_NAME = "Bigtable";
   private static final String TRACING_INNER_CLIENT_NAME = "BaseBigtable";
+  private static final MutateRowsBatchingDescriptorV2 batchingDescriptorV2 =
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: is this going to live side by side with MutateRowsBatchingDescriptor? Can we replace the old one with this?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+/**
+ * A custom implementation of a {@link BatchingDescriptor} to split Rpc response into individual
+ * rows response and in a {@link MutateRowsException}.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications directly.
+ */
+// TODO(rahulkql): Consider renaming this class
+public class MutateRowsBatchingDescriptorV2
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:39]: can we remove "public" if it is internal only? If no, let's add `@InternalApi`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java (hunk 7 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Thanks for the review.
Yes, We are aiming to remove existing `MutateRowsBatchingDescriptor`. I have added a TODO in `MutateRowsBatchingDescriptorV2` for the same reason.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java (hunk 1 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: I have kept the method description Shall we remove those as well?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java (hunk 2 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java:116]: Thanks for pointing it out.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 5 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Actually, I mistakenly assumed that we should not deprecate and change the Request wrapper RowMutation to BulkMutation in one go.

But I have updated it after you pointed it out.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 6 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Removed the same.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java (hunk 8 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java:39]: We would need this class in other packages so I have marked it as `@InternalApi("For internal use only")`.

**Comment by rahulKQL** [comment]
@igorbernstein2 @kolea2 
My apologies for the late response, I have addressed the feedback comments. 

Could you please have a fresh look at this PR whenever have time.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java (hunk 1 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: yes please

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 9]
```diff
@@ -921,6 +909,34 @@ public void bulkMutateRows(BulkMutation mutation) {
...
+  /**
+   * Mutates multiple rows in a batch. Each individual row is mutated atomically as in MutateRow,
+   * but the entire batch is not executed atomically.
+   *
+   * <p>Sample Code:
+   *
+   * <pre>{@code
+   * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create("[PROJECT]", "[INSTANCE]")) {
+   *   try (Batcher<RowMutationEntry, Void> batcher = bigtableDataClient.newBulkMutationBatcher("[TABLE]")) {
+   *     for (String someVale : someCollections) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: `someValue : someCollection`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java [hunk 10]
```diff
@@ -34,7 +34,7 @@
  * <p>This class is meant for manual batching, while {@link BulkMutationBatcher} is meant for
  * automatic batching with flow control.
  */
-public final class BulkMutation implements Serializable {
+public final class BulkMutation implements Serializable, Cloneable {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java:37]: Please add a test to make sure that clone doesn't share data with the original and has all of the necessary attributes

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+package com.google.cloud.bigtable.data.v2.models;
+
+import com.google.api.core.InternalApi;
+import com.google.bigtable.v2.MutateRowsRequest;
+import com.google.protobuf.ByteString;
+import java.io.Serializable;
+import javax.annotation.Nonnull;
+
+/**
+ * Represents a list of mutations targeted at a single row. It's meant to be used as an parameter
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: Please add a comment that the mutations will be applied atomically and in order

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    return create(ByteString.copyFromUtf8(key));
+  }
+
+  /** Creates a new instance of the mutation builder. */
+  public static RowMutationEntry create(@Nonnull ByteString key) {
+    return new RowMutationEntry(key, Mutation.create());
+  }
+
+  /** Creates a new instance of the mutation builder. */
+  public static RowMutationEntry create(@Nonnull String key, @Nonnull Mutation mutation) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: Is it necessary to expose this?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    return create(ByteString.copyFromUtf8(key), mutation);
+  }
+
+  /** Creates a new instance of the mutation builder. */
+  public static RowMutationEntry create(@Nonnull ByteString key, @Nonnull Mutation mutation) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: Is it necessary to expose this?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java [hunk 14]
```diff
@@ -477,7 +495,10 @@ public static EnhancedBigtableStub create(EnhancedBigtableStubSettings settings)
   /**
    * Returns the callable chain created in {@link #createBulkMutateRowsBatchingCallable()} ()}
    * during stub construction.
+   *
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStub.java]: Please remove doc except for the deprecation notice.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java (hunk 12 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: I don't think it is necessary, I copied these to have consistency between `RowMutation`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+      }
+    }
+  }
+
+  /** {@inheritDoc} */
+  public long countBytes(RowMutationEntry request) {
+    return request.toProto().getSerializedSize();
+  }
+
+  /** A {@link BatchingRequestBuilder} that can aggregate BulkMutation */
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: that will spool mutations and send them out as a `{@link BulkMutation}`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+                return input.getIndex();
+              }
+            });
+
+    int i = 0;
+    for (SettableApiFuture<Void> response : batch) {
+      // NOTE: The gax batching api doesn't allow for a single response to get different
+      // exception for different entries. However this does not affect this client because
+      // SettableApiFuture corresponds to  single element. So just use the last error per entry.
+      ApiException lastError = null;
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: I dont understand what you mean here. How can a single entry get multiple errors?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+    MyRequestBuilder(BulkMutation prototype) {
+      this.bulkMutation = prototype.clone();
+    }
+
+    @Override
+    public void add(RowMutationEntry rowEntry) {
+      MutateRowsRequest.Entry entry = rowEntry.toProto();
+      ByteString rowK = entry.getRowKey();
+      bulkMutation.add(rowK, Mutation.fromProtoUnsafe(entry.getMutationsList()));
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: Can BulkMutation take an instance of RowMutationEntry?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+      if (lastError == null) {
+        response.set(null);
+      } else {
+        response.setException(lastError);
+      }
+    }
+  }
+
+  /** {@inheritDoc} */
+  public long countBytes(RowMutationEntry request) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: s/request/entry

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+  /** A {@link BatchingRequestBuilder} that can aggregate BulkMutation */
+  static class MyRequestBuilder implements BatchingRequestBuilder<RowMutationEntry, BulkMutation> {
+    private BulkMutation bulkMutation;
+
+    MyRequestBuilder(BulkMutation prototype) {
+      this.bulkMutation = prototype.clone();
+    }
+
+    @Override
+    public void add(RowMutationEntry rowEntry) {
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: Please be consistent in your variable names

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/RowMutationEntryBatcherIT.java [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+                  RowCell.create(
+                      family,
+                      ByteString.copyFromUtf8("qualifier"),
+                      10_000L,
+                      Lists.<String>newArrayList(),
+                      ByteString.copyFromUtf8("value-" + i)))));
+    }
+    ServerStream<Row> actualRows = client.readRows(Query.create(tableId).prefix(rowPrefix));
+
+    Truth.assertThat(actualRows).containsExactlyElementsIn(expectedRows);
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/RowMutationEntryBatcherIT.java]: static import please

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java [hunk 21]
```diff
@@ -418,7 +420,7 @@ public void bulkMutateRowsSettingsAreNotLostTest() {
     assertThat(builder.bulkMutateRowsSettings().getRetryableCodes())
         .containsAtLeast(Code.ABORTED, Code.DEADLINE_EXCEEDED);
     assertThat(builder.bulkMutateRowsSettings().getRetrySettings()).isEqualTo(retrySettings);
-    assertThat(builder.bulkMutateRowsSettings().getBatchingSettings())
+    assertThat(builder.bulkMutateRowsSettings().build().getBatchingSettings())
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java]: why is this change necessary?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java (hunk 12 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: I think those were added to RowMutation for explicit usage by the hbase client

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java (hunk 21 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java]: We don't have getters anymore in new BatchingCallSettings.Builder. So I had an option either to review this assertion or build and get the BatchingSettings.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java (hunk 16 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: I was trying to map FailedMutation to each entry, I had assumed that `MutateRowsException#getFailedMutation()` could return different number of FailedMutation than the number of entries on that batch.

In case the number of entries exceeds the number of FailedMutation, we are setting the last failedMutation to the rest of the entries.

Should we set throwable to each entry (without checking MutateRowsException)? Something like this:
```java
  ...
  @Override
  public void splitException(Throwable throwable, List<SettableApiFuture<Void>> batch) {
    for (SettableApiFuture<Void> future : batch) {
      future.setException(throwable);
    }
  }
  ...
```

Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: The first approach is correct (unpacking the entry errors from `MutateRowsException#getFailedMutations`). However the comment is misleading. Each entry will have at most 1 error. In fact if there were duplicate keys then `Maps.uniqueIndex` would throw an IllegalArgumentException. To be honest I think this would be easier to read:

```java
Map<Integer, Throwable> entryErrors = new HashMap<>();
for(FailedMutation failure : failedMutations) {
  entryErrors.put(failure.getIndex(), failure.getError();
}

int i = 0;
for (SettableApiFuture<Void> entryResultFuture : batch) {
  Throwable entryError = entryErrors.get(i++);
  if (entryError == null) {
     entryResultFuture.set(null);
  } else {
     entryResultFuture.setException(entryError);
  }
}    
```

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java (hunk 21 above)
Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java]: I would just add a getter to BatchingCallSettings

Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java]: Did you mean add a getter to `BatchingCallSettings.Builder`?

Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettingsTest.java]: yep

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java (hunk 10 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutation.java:37]: I have added a test case for this in `BulkMutation#cloneTests()`.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java (hunk 12 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/RowMutationEntry.java]: Thanks for clarifying it, I have removed those from `RowMutationEntry`.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java (hunk 16 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: Thanks a lot for simplifying it. I have updated this method with the same approach.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java (hunk 17 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: Added a `BulkMutation#add(RowMutationEntry)`, I hope that should be fine.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java (hunk 19 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: Thanks, Will now onwards try to be consistent.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+  @Override
+  public long countBytes(RowMutationEntry entry) {
+    return entry.toProto().getSerializedSize();
+  }
+
+  /**
+   * A {@link BatchingRequestBuilder} that will spool mutations and send them out as a {@link
+   * BulkMutation}.
+   */
+  static class MyRequestBuilder implements BatchingRequestBuilder<RowMutationEntry, BulkMutation> {
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/mutaterows/MutateRowsBatchingDescriptorV2.java]: minor - let's rename this. RequestBuilder maybe?

**Comment by rahulKQL** [comment]
Closing this PR as this is already merged in origin repo.
